protocol/network/7-properties.md

```diff
@@ -1,16 +1,105 @@
+7 Properties
+============
 
-# IPFS Protocol on the Wire
+## 7.1 Communication Model - Streams
 
-Authors: [Juan Benet](http://github.com/jbenet)
+The Network layer handles all the problems of connecting to a peer, and exposes
+simple bidirectional streams. Users can both open a new stream
+(`NewStream()`) and register a stream handler (`SetStreamHandler`). The user
+is then free to implement whatever wire messaging protocol she desires. This
+makes it easy to build peer-to-peer protocols, as the complexities of
+connectivity, multi-transport support, flow control, and so on, are handled.
```

Comment:
btw, if we _don't need_ the distinction between the dialer/listener it may be nice to just think of `NewStream`. but i'm not sure. im not sure we can completely remove directionality. it would sort of look like a TCP simultaneous open...

for example
```go
// in one side
n := p2p.NewNode(config)
p := n.Mount("/dht", DialAndListen)
s1, err := p.NewStream()
// s1 above and s2 here same pipe
```

```
// in the other side
n := p2p.NewNode(config)
p := n.Mount("/dht", DialAndListen)
s2, err := p.NewStream()
// s1 above and s2 here same pipe
```
- We would need some material to distinguish streams and imply that we do want simultaneous open to happen, or we want _two separate_ duplex pipes.
- It may be impossible to conveniently distinguish these.
- on the other hand, we may want to just leave that up to the user...

### \o/ another reroll \o/ 

**actually i think this is wrong and likely very confusing. interesting though. i'll leave it here for posterity. but not a serious suggestion:**

```go
n := libp2p.NewNode(...all the config...)
stream1, err := p.NewStream("/dht/1") // returns a stream mounted at /dht/1. other side can simultaneous-open to this one.

stream2, err := n.DialStream("/dht/0") // assigns next available "port number", dials out.
stream3, err := n.AcceptStream() // assigns next available "port number" (0 is implied).

// same as
p := n.Mount("/dht", DialAndListen)
s1, err := p.Dial("1")
s2, err := p.Dial("0") // assigns next port number
s3, err := p.Accept() // "0
```